scilpy/tractanalysis/features.py

```diff
@@ -0,0 +1,162 @@
+import numpy as np
+
+
+def remove_loops_and_sharp_turns(streamlines,
+                                 max_angle,
+                                 use_qb=False,
+                                 qb_threshold=15.):
+    """
+    Remove loops and sharp turns from a list of streamlines.
+    Parameters
+    ----------
+    streamlines: list of ndarray
+        The list of streamlines from which to remove loops and sharp turns.
+    use_qb: bool
+        Set to True if the additional QuickBundles pass is done.
+        This will help remove sharp turns. Should only be used on
+        bundled streamlines, not on whole-brain tractograms.
+    max_angle: float
+        Maximal winding angle a streamline can have before
+        being classified as a loop.
+    qb_threshold: float
+        Quickbundles distance threshold, only used if use_qb is True.
+    Returns
+    -------
+    A tuple containing
+        list of ndarray: the clean streamlines
+        list of ndarray: the list of removed streamlines, if any
+    """
+
+    loops = []
+    streamlines_clean = []
+    for s in streamlines:
+        if tm.winding(s) >= max_angle:
+            loops.append(s)
+        else:
+            streamlines_clean.append(s)
+
+    # TODO can we use QBx instead?
+    if use_qb:
+        if len(streamlines_clean) > 1:
+            streamlines = streamlines_clean
+            curvature = []
+            streamlines_clean = []
+
+            qb = QuickBundles(threshold=qb_threshold)
+            clusters = qb.cluster(streamlines)
+
+            for cc in clusters.centroids:
+                curvature.append(tm.mean_curvature(cc))
+            mean_curvature = sum(curvature)/len(curvature)
+
+            for i in range(len(clusters.centroids)):
+                if tm.mean_curvature(clusters.centroids[i]) > mean_curvature:
+                    for indice in clusters[i].indices:
+                        loops.append(streamlines[indice])
+                else:
+                    for indice in clusters[i].indices:
+                        streamlines_clean.append(streamlines[indice])
+        else:
+            logging.debug("Impossible to use the use_qb option because " +
+                          "not more than one streamline left from the\n" +
+                          "input file.")
+
+    return streamlines_clean, loops
+
+
+def get_streamlines_bounding_box(streamlines):
+    box_min = np.array([np.inf, np.inf, np.inf])
+    box_max = -np.array([np.inf, np.inf, np.inf])
+
+    for s in streamlines:
+        box_min = np.minimum(box_min, np.min(s, axis=0))
+        box_max = np.maximum(box_max, np.max(s, axis=0))
+
+    return box_min, box_max
+
+
+def prune(streamlines, threshold, features):
+    indices = np.arange(len(streamlines))
+
+    outlier_indices = indices[features < threshold]
+    rest_indices = indices[features >= threshold]
+
+    return outlier_indices, rest_indices
+
+
+# TODO could replace QB by QBx. Would need to adjust thresholds.
```

Comment:
Same comment here

Reply:
Same note :)

Reply:
Do we really want to incorporate hierarchical clustering in a function that does hierarchical clustering ? I think that would mess with the intend of Marc-Alex to simply evaluate QB at different layer.

Reply:
The QBx would replace the hierarchical clustering ;) But we tried this internally, and needed to adjust some thresholds. I see that you removed the comment. That's also good for me.